vscode_projects/__init__.py

```diff
@@ -0,0 +1,545 @@
+            if foundPM:
+                self._projectManagerEnabled = True
+                self.writeConfig("projectManagerEnabled", True)
+            else:
+                self._projectManagerEnabled = False
+        else:
+            self._projectManagerEnabled = projectManagerEnabled
+
+        # Priority settings
+        for p in self._sortPriority:
+            prio = self.readConfig(f"priority{p}", int)
+            if prio is None:
+                self.writeConfig(f"priority{p}", self._sortPriority[p])
+            else:
+                self._sortPriority[p] = prio
+
+        # Terminal command setting
+        terminalCommand = self.readConfig('terminalCommand', str)
+        if terminalCommand is not None:
+            self._terminalCommand = terminalCommand
+
+    # Strings are normalized to match without accents and casing
+    def _normalizeString(self, input: str) -> str:
+        return ''.join(c for c in unicodedata.normalize('NFD', input)
+                       if unicodedata.category(c) != 'Mn').lower()
+
+    def handleTriggerQuery(self, query):
+        if not query.isValid:
+            return
+
+        if query.string == "":
+            return
+
+        matcher = Matcher(query.string)
+
+        results: dict[str, SearchResult] = {}
+
+        if self.recentEnabled:
+            results = self._searchInRecentFiles(matcher, results)
+
+        if self.projectManagerEnabled:
+            results = self._searchInProjectManager(matcher, results)
+
+        sortedItems = sorted(results.values(), key=lambda item: "%s_%s_%s" % (
+            '{:03d}'.format(item.priority), '{:03d}'.format(item.sortIndex), item.project.name), reverse=False)
+
+        for i in sortedItems:
+            query.add(self._createItem(i.project, query))
+
+    # Creates an item for the query based on the project and plugin settings
+    def _createItem(self, project: Project, query: Query) -> StandardItem:
+        actions: list[Action] = []
+
+        if self.terminalCommand != "":
+            actions.append(
+                Action(
+                    id="open-terminal",
+                    text=f"Run terminal command in project's workdir: {self.terminalCommand}",
+                    callable=lambda: runTerminal(
+                        close_on_exit=True,
+                        script=self.terminalCommand,
+                        workdir=project.path,
+                    )
+                )
+            )
+
+        actions.append(
+            Action(
+                id="open-code",
+                text="Open with VSCode",
+                callable=lambda: runDetachedProcess(
+                    ["code", project.path]),
+            )
+        )
+
+        return StandardItem(
+            id=project.path,
+            text=project.displayName,
+            subtext=project.path,
+            iconUrls=[f"file://{Path(__file__).parent}/icon.svg"],
```

Comment:
what if you remove the two slashes?

Reply:
Leaving it there was an oversight, will push the commit momentarily, I was actually testing multiple cases and in my case `file:...` works, `file://...` works as well and suprisingly any amount of slashes works too.
I was refering to the [initialization](https://github.com/albertlauncher/python/pull/198/files#diff-3c8b5db043c3c17ea3751645c9353138203d2aa594c8a67eafdf994ee47b381bR163), that's where it stopped working. It stopped working for all python plugins, it shows some kind of fallback symbol instead:

![image](https://github.com/albertlauncher/python/assets/17074367/61830607-4357-4338-87ba-af11d6b9d298)

Reply:
Just to clarify, it works fine for the StandardItem class:
![image](https://github.com/albertlauncher/python/assets/17074367/b49efa66-6b22-4a14-8563-f42f485f03ba)

Reply:
The mpris icon problem looks like a font config issue. The icons there use emojis.

Reply:
![IMAGE 2024-06-30 13:09:28](https://github.com/albertlauncher/python/assets/3252315/3edb1d5f-d99b-4c07-9baf-9573f2cf81ac)

this is what it usually looks like (this is on vanilla manjaro)

Reply:
Ah okay, that makes sense, so where would the `self.iconUrls` be shown? Maybe a better example is the [iconUrls in wikipedia plugin](https://github.com/albertlauncher/python/blob/main/wikipedia/__init__.py#L28)

When I enter `wiki`, I get the following results:

![image](https://github.com/albertlauncher/python/assets/17074367/d3ac2f88-7259-4e85-9fe3-ec2faaddc1a8)

I would assume the first icon would use the iconUrls, but seems to fallback to some kind of emoji that my system font cannot render (my problem, not blaming anyone:)
The second one I would understand uses the "puzzle" emoji (just like mpris configuration row from your screenshot)

Though when I enter `wiki something` I get the iconUrl, but that is already rendered through StandardItem result

![image](https://github.com/albertlauncher/python/assets/17074367/412b49c8-b8e8-4504-9d16-db682a76e3af)

Reply:
These items are not from the wiki query handler. The first one is the built in TriggersQueryHandler and the second from the pluginqueryhandler

Reply:
Understood, so is there any point specifying the iconUrls on the plugin itself? Searching through the source code, I don't really see many references to that except for FallbacksModel, but I'm not exactly sure what that is used for. But majority of python plugins still define it on the plugin instance.

Reply:
Yes, to save reasources and fast query times. The icons stored in the plugin instance are static, so there is no need to build a path and list on every query for each item.

Reply:
> The icons stored in the plugin instance are static, so there is no need to build a path and list on every query for each item

Understood, I'll make the changes as appropriate, my mistake

> Do you struggle to understand the API? If so do you know the stubfile (albert.pyi)?

I think I just did not connect the dots, I barely use python and the whole concept of decorators for properties and such is a little bit foreign to me, so originally (a few years ago) I just modified the jetbrains plugin for vscode. Nowdays I'm a little bit more familiar with both the language and the API, but the I totally missed the path caching for some reason.

> I dont get that. I dont think the fallbacksmodel uses plugininstance fields.

No, that was due to my misunderstanding of iconUrls usage on the plugin instance. I was trying to find references where it could be used in the albert core. Now it's clear to me that it's just internally used inside the plugins themselves.

Sorry for the confusion and thanks for the patience :)